http/routing.hpp

```diff
@@ -1299,28 +1311,139 @@ class Router
+                               " local user but is missing or wrong type";
+                        asyncResp->res.result(
+                            boost::beast::http::status::internal_server_error);
+                        return;
+                    }
+                }
+
+                // Get the userprivileges from the role
+                redfish::Privileges userPrivileges =
+                    redfish::getUserPrivileges(userRole);
+
+                // Set isConfigureSelfOnly based on D-Bus results.  This
+                // ignores the results from both pamAuthenticateUser and the
+                // value from any previous use of this session.
+                req.session->isConfigureSelfOnly = passwordExpired;
+
+                // Modifyprivileges if isConfigureSelfOnly.
+                if (req.session->isConfigureSelfOnly)
+                {
+                    // Remove allprivileges except ConfigureSelf
+                    userPrivileges = userPrivileges.intersection(
+                        redfish::Privileges{"ConfigureSelf"});
+                    BMCWEB_LOG_DEBUG << "Operation limited to ConfigureSelf";
+                }
+
+                if (!rules[ruleIndex]->checkPrivileges(userPrivileges))
+                {
+                    asyncResp->res.result(
+                        boost::beast::http::status::forbidden);
+                    if (req.session->isConfigureSelfOnly)
+                    {
+                        redfish::messages::passwordChangeRequired(
+                            asyncResp->res,
+                            "/redfish/v1/AccountService/Accounts/" +
+                                req.session->username);
+                    }
+                    return;
+                }
+
+                req.userRole = userRole;
+
+                // any uncaught exceptions become 500s
+                try
+                {
+                    rules[ruleIndex]->handleUpgrade(
+                        req, asyncResp, std::forward<Adaptor>(adaptor));
+                }
+                catch (const std::exception& e)
+                {
+                    BMCWEB_LOG_ERROR << "An uncaught exception occurred: "
+                                     << e.what();
+                    asyncResp->res.result(
+                        boost::beast::http::status::internal_server_error);
+                    return;
+                }
+                catch (...)
+                {
+                    BMCWEB_LOG_ERROR << "An uncaught exception occurred. The "
+                                        "type was unknown "
+                                        "so no information was available.";
+                    asyncResp->res.result(
+                        boost::beast::http::status::internal_server_error);
+                    return;
+                }
+            },
+            "xyz.openbmc_project.User.Manager", "/xyz/openbmc_project/user",
```

Comment:
Still comments upstream about this being a blocking call and such.. 
https://gerrit.openbmc-project.xyz/c/openbmc/bmcweb/+/46991/